cartographer/mapping_2d/sparse_pose_graph.cc

```diff
@@ -427,6 +427,14 @@ void SparsePoseGraph::AddNodeFromProto(const transform::Rigid3d& global_pose,
   });
 }
 
+void SparsePoseGraph::AddNodeToSubmap(const mapping::NodeId& node_id,
+                                      const mapping::SubmapId& submap_id) {
+  common::MutexLocker locker(&mutex_);
+  AddWorkItem([this, node_id, submap_id]() REQUIRES(mutex_) {
+    submap_data_.at(submap_id).node_ids.insert(node_id);
```

Comment:
Just to make you aware, this has already been [implemented in `AddSerializedConstraints`](https://github.com/googlecartographer/cartographer/commit/7964211fef53b13831a09baaa022fdce80f6c19e#diff-6d03d3eb04a3406b2be0eda40f99e843R454), although `AddSerializedConstraints` is not used yet anywhere.

It's planned that `AddSerializedConstraints` will be called from `LoadMap` only if `load_frozen = false`, as you have suggested earlier. However, you want to call `AddNodeToSubmap` unconditionally. Do you have thoughts on this?

Reply:
This is best discussed in the context of the PR that makes use of `AddSerializedConstraints()`. I'm not too sure there is much to deduplicate though. For loading a frozen map (and not continuing SLAM) we do not want to add constraints, but the information which nodes are part of which submap is still needed.